Update and expand gpu instance support.

nuclearsandwich opened the pull request:
This PR started because we were trying to use the g5.xlarge instances which had slightly different lspci output from what the current detector expected. In the end I also added support for different nvidia drivers depending on the detected hardware as the old driver required for supporting optimus was too old to support the hardware in g4n and g5 AWS instances.

As part of these changes, I also moved GPU detection into a plugin for chef's ohai utlity, which profiles a system on chef startup so its attributes are available to chef cookbooks, so I could use more information about these GPUs in other parts of the cookbook beyond only_if and not_if predicates.

---

The g5.xlarge instances we are trying to deploy have the following lspci output:

```
00:00.0 Host bridge: Intel Corporation 440FX - 82441FX PMC [Natoma]
00:01.0 ISA bridge: Intel Corporation 82371SB PIIX3 ISA [Natoma/Triton II]
00:01.3 Non-VGA unclassified device: Intel Corporation 82371AB/EB/MB PIIX4 ACPI (rev 08)
00:03.0 VGA compatible controller: Amazon.com, Inc. Device 1111
00:04.0 Non-Volatile memory controller: Amazon.com, Inc. Device 8061
00:05.0 Ethernet controller: Amazon.com, Inc. Elastic Network Adapter (ENA)
00:1e.0 3D controller: NVIDIA Corporation Device 2237 (rev a1)
00:1f.0 Non-Volatile memory controller: Amazon.com, Inc. NVMe SSD Controller
 ```

(Note the use of 3D controller rather than VGA compatible controller)
This updates the nvidia detection to try and use this.

Since this does not match the nvidia grid output from the existing AWS
support I did not update that to use it.

Further iteration may be required to get these working.

nuclearsandwich committed:
Include g5.xlarge instances in generic GPU detection.

The g5.xlarge instances we are trying to deploy have the following lspci
output:

```
00:00.0 Host bridge: Intel Corporation 440FX - 82441FX PMC [Natoma]
00:01.0 ISA bridge: Intel Corporation 82371SB PIIX3 ISA [Natoma/Triton II]
00:01.3 Non-VGA unclassified device: Intel Corporation 82371AB/EB/MB PIIX4 ACPI (rev 08)
00:03.0 VGA compatible controller: Amazon.com, Inc. Device 1111
00:04.0 Non-Volatile memory controller: Amazon.com, Inc. Device 8061
00:05.0 Ethernet controller: Amazon.com, Inc. Elastic Network Adapter (ENA)
00:1e.0 3D controller: NVIDIA Corporation Device 2237 (rev a1)
00:1f.0 Non-Volatile memory controller: Amazon.com, Inc. NVMe SSD Controller
 ```

(Note the use of 3D controller rather than VGA compatible controller)
This updates the nvidia detection to try and use this.

Since this does not match the nvidia grid output from the existing AWS
support I did not update that to use it.

Further iteration may be required to get these working.

Crola1702 reviewed (approved):
LGTM

> Further iteration may be required to get these working.

Is this because our scripts work with VGA Nvidia controller?

nuclearsandwich committed:
Add a cookbook ohai plugin for doing GPU detection in ohai.

We have some helpers shelling out to lspci during the cookbook run,
however, as we extend the amount of information we're getting from lspci
we also need to use it in more places. Getting the information in ohai
means that we can use it via node attributes all around our chef recipes
in a consistent fashion.

nuclearsandwich committed:
Add helper for getting nvidia devices from ohai plugin.

nuclearsandwich committed:
Refactor existing helpers to use nvidia_devices method.

nuclearsandwich committed:
Install different nvidia drivers depending on hardware available.

nuclearsandwich committed:
Handle an empty list of nvidia devices better.

When run on a host without a detected nvidia driver this logic wasn't
sufficient.

nuclearsandwich committed:
Make compatible optional.

A typo when this regex was transcribed prevented it from being used to
correctly detect `3D Controller`.

nuclearsandwich committed:
Fix incorrect return value for nvidia_devices.

I neglected to convert the hash structure from node attributes into an
array as documented. There is no loss of information in doing so since
the key for each device is also stored as the `slot` value.

nuclearsandwich committed:
Add GPU case for AWS g5.xlarge.

nuclearsandwich committed:
Add support for g4dn instances.

g4dn are an Nvidia-based accelerated compute instance and are even
cheaper still than g3s right now.

nuclearsandwich commented on ohai/gpu_devices.rb:
```diff
@@ -0,0 +1,24 @@
+    shell_out('lspci -vmm').stdout.split("\n\n").select do |pcidev|
+      pcidev =~ /(?:3D|VGA)(?: compatible)? controller/
+    end.map do |gpudev|
+      Mash.new.tap do |properties|
+        gpudev.lines.map do |line|
+          line.chomp!
+          key, val = line.split(":\t")
+          properties[key.downcase] = val
+        end
+      end
```
Ohai attributes are generally all object/record/hash/dict formatted rather than list/array formatted so that different data sources can uniquely identify each attribute. As a result, I'm storing the gpu_devices not as a list, but as a hash (using Ohai's Mash class which is just a fancy Hash) with the `slot` value as the key, since this is the only attribute guaranteed to be unique on a system (no two PCI devices can occupy the same slot, I hope) whereas it's possible to have multiple identical GPUs on a single host so I could not use the `device` field.

nuclearsandwich commented on ohai/gpu_devices.rb:
```diff
@@ -0,0 +1,24 @@
+  provides 'gpu_devices'
+
+  def gpu_devices_from_lspci
+    shell_out('lspci -vmm').stdout.split("\n\n").select do |pcidev|
+      pcidev =~ /(?:3D|VGA)(?: compatible)? controller/
+    end.map do |gpudev|
+      Mash.new.tap do |properties|
+        gpudev.lines.map do |line|
+          line.chomp!
+          key, val = line.split(":\t")
+          properties[key.downcase] = val
+        end
+      end
+    end
```
This is pretty magical ruby, but Ohai makes it somewhat hard to avoid.

The output from lspci's verbose, modern, machine-readable (`-vmm`) is in not-quite-deb822 formatted stanzas, one per device. I split on double newlines to break up the device stanzas and then filter out the ones which match a GPU device regex.

The regex was crafted based on existing observations of various hosts and instance types. This would technically match `3D compatible controller` which is not a class I've seen in the wild but would probably still be a GPU anyway, so I'm not bothered.

The individual key value lines are then split by the separator string `:<tab>` and then placed in a new Mash containing the properties for that gpu device. The `tap` method is a common Ruby idiom which returns the tapped object after manipulating it within the block. The use of this idiom just prevents accidentally returning some other value to the outer `map` block instead of the newly created Mash.

nuclearsandwich commented on ohai/gpu_devices.rb:
```diff
@@ -0,0 +1,24 @@
+  end
+
+  collect_data(:linux) do
+    gpu_devices Mash.new
+    gpu_devices_from_lspci.each do |dev|
+      gpu_devices[dev['slot']] = dev
+    end
```
Ohai attributes are generally all object/record/hash/dict formatted rather than list/array formatted so that different data sources can uniquely identify each attribute. As a result, I'm storing the gpu_devices not as a list, but as a hash (using Ohai's Mash class which is just a fancy Hash) with the `slot` value as the key, since this is the only attribute guaranteed to be unique on a system (no two PCI devices can occupy the same slot, I hope) whereas it's possible to have multiple identical GPUs on a single host so I could not use the `device` field.

nuclearsandwich commented on recipes/agent_build_tools.rb:
```diff
@@ -52,8 +52,44 @@
+end
+nvidia_device = if has_nvidia_support?
+                  nvidia_devices.first['device']
+                else
+                  nil
+                end
```
An empty `else` clause would technically evaluate to `nil` anyway but since we're using it as a sigil value below I thought it best to be explicit.

nuclearsandwich commented on recipes/agent_build_tools.rb:
```diff
@@ -52,8 +52,44 @@
+                  # if there is no nvidia device but specify
+                  # one to avoid a warning.
+                  'nvidia-driver-470'
+
```
```suggestion
```

nuclearsandwich committed:
Remove spurious newline.

nuclearsandwich commented on recipes/agent_build_tools.rb:
```diff
@@ -52,8 +52,43 @@
+                  nil
+                end
+
+nvidia_driver = case nvidia_device
+                when /GTX 550/
+                  # Old optimus machine in OSRF office
+                  'nvidia-384'
+                when /GRID K520/
+                  # AWS g2 instances
+                  'nvidia-384'
+                when /Tesla M60/
+                  # AWS g3 instances
+                  'nvidia-driver-470'
+                when /Tesla T4/
+                  # AWS g4dn instances
+                  'nvidia-driver-470'
+                when /Device 2237/
+                  # AWS g5 instances
+                  'nvidia-driver-470'
+                when nil
+                  # it doesn't really matter which driver we use
+                  # if there is no nvidia device but specify
+                  # one to avoid a warning.
+                  'nvidia-driver-470'
+                else
+                  Chef::Log.warn("Untested GPU `#{nvidia_devices.first['device']}` being used. Assuming a functioning driver")
+                  'nvidia-driver-470'
+                end
```
If this grows much more it may make sense to move it into a helper method, especially if different considerations like platform start being considered as well.

nuclearsandwich commented:
Although @Crola1702's review is slightly stale. I am proceeding to merge this PR since I now have machines on the production build farm using these changes! I'll commit to addressing any after-the-fact feedback when I present the changes at the next infra meeting.
